examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+;;;;
+;;;;
+;;;; A Brainfold/Brainf*** compiler implemented in Coalton
+;;;;
+;;;; The compiler handles all standard bf commands:
+;;;; ( > < + - . , [ ] ).
+;;;;
+;;;;
+;;;; Run Brainfold programs with (run-program "+++++[-]+++")
+;;;;
+;;;; or
+;;;;
+;;;; (run-file "/path/to/your/file.bf")
+;;;;
+;;;; try (coalton (hello-world)) in the REPL!
+;;;;
+
+
+(cl:defpackage #:brainfold
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+   (#:vec #:coalton-library/vector)
+   (#:iter #:coalton-library/iterator)
+   (#:cell #:coalton-library/cell)
+   (#:char #:coalton-library/char)
+   (#:str #:coalton-library/string)
+   (#:list #:coalton-library/list)
+   (#:arith #:coalton-library/math))
+
+  (:export
+   #:eval
+   #:run-program
+   #:run-file
+   #:hello-world))
+
+;;;
+;;; Environment (Vector, Pointer, Print Buffer)
+;;;
+
+(in-package #:brainfold)
+
+(named-readtables:in-readtable coalton:coalton)
+
+(coalton-toplevel
+
+  ;;
+  ;; Initializing a vector with 0's, ideally 30k according to the
+  ;; specifications
+  
+  (declare bf-array-size (UFix))
+  (define bf-array-size 1000)
+
+  (define (generate-bf-array)
+    "Initializes the brainfold array."
+    (let v = (vec:new))
+    (vec:extend! v (iter:repeat-for 0 bf-array-size))
+    v)
+ 
+  (define bf-array (generate-bf-array))
+
+  (define (reset-bf-array)
+    "Resets the brainfold array."
+    (vec:clear! bf-array)
+    (vec:extend! bf-array (iter:repeat-for 0 bf-array-size))
+    bf-array)
+
+  ;;
+  ;; The pointer to the current Brainfold cell
+  ;;
+  
+  (declare pointer (Cell UFix))
+  (define pointer (cell:new 0))
+
+  ;;
+  ;; Accessing Brainfold cells using pointer
+  ;;
+  
+  (define (nth-bf-cell index)
+    "Finds the vector entry associated with the given index."
+    (vec:index index bf-array))
+
+  (define (current-bf-cell-value)
+    "Returns the value of the bf-cell indexed by the pointer."
+    (unwrap (nth-bf-cell (cell:read pointer))))
+
+  ;;
+  ;; Print Buffer for printing
+  ;;
+
+  (define print-buffer (cell:new ""))
+  
+  ;;
+  ;; Clean Environment
+  ;;
+  
+  (define (reset-bf)
+    "Resets the bf-environment- pointer, print-buffer, and array."
+    (cell:write! pointer 0)
+    (cell:write! print-buffer "")
+    (reset-bf-array)))
+
+;;;
+;;; Commands (Functions called by Brainfold Cmds)
+;;;
+
+(coalton-toplevel
+
+  ;;
+  ;; Navigating through bf-cells (> <)
+  ;;
+  
+  (define (move-right)
```

Comment:
One elegant thing about having a `State` type is that _all_ instructions become functions mapping us from `State` to `State`. If you ever decide to rewrite your data structure to be immutable, you would be able to do it easily.